source/common/http/conn_pool_grid.cc

```diff
@@ -0,0 +1,185 @@
+  // to the active pool.
+  cancellable_->cancel(cancel_policy);
+  delete this;
+}
+
+ConnectivityGrid::ConnectivityGrid(
+    Event::Dispatcher& dispatcher, Random::RandomGenerator& random_generator,
+    Upstream::HostConstSharedPtr host, Upstream::ResourcePriority priority,
+    const Network::ConnectionSocket::OptionsSharedPtr& options,
+    const Network::TransportSocketOptionsSharedPtr& transport_socket_options,
+    Upstream::ClusterConnectivityState& state, TimeSource& time_source,
+    ConnectivityOptions connectivity_options)
+    : dispatcher_(dispatcher), random_generator_(random_generator), host_(host),
+      priority_(priority), options_(options), transport_socket_options_(transport_socket_options),
+      state_(state), time_source_(time_source) {
+  // TODO(#15649) support v6/v4, WiFi/cellular.
+  ASSERT(connectivity_options.protocols_.size() == 3);
+  ASSERT(contains(connectivity_options.protocols_,
+                  {Http::Protocol::Http11, Http::Protocol::Http2, Http::Protocol::Http3}));
+}
+
+ConnectivityGrid::~ConnectivityGrid() {
+  // Ignore drained callbacks during teardown.
+  destroying_ = true;
+}
+
+absl::optional<ConnectivityGrid::PoolIterator> ConnectivityGrid::createNextPool() {
+  // Pools are created by newStream, which should not be called during draining.
+  ASSERT(drained_callbacks_.empty());
+  // Right now, only H3 and ALPN are supported, so if there are 2 pools we're done.
+  if (pools_.size() == 2 || !drained_callbacks_.empty()) {
+    return absl::nullopt;
+  }
+
+  // HTTP/3 is hard-coded as higher priority, H2 as secondary.
+  if (pools_.empty()) {
+    pools_.push_back(Http3::allocateConnPool(dispatcher_, random_generator_, host_, priority_,
+                                             options_, transport_socket_options_, state_,
+                                             time_source_));
+    return pools_.begin();
+  }
+  pools_.push_back(std::make_unique<HttpConnPoolImplMixed>(dispatcher_, random_generator_, host_,
+                                                           priority_, options_,
+                                                           transport_socket_options_, state_));
+  return pools_.begin()++;
+}
+
+bool ConnectivityGrid::hasActiveConnections() const {
+  // This is O(n) but the function is constant and there are no plans for n > 8.
+  for (auto& pool : pools_) {
+    if (pool->hasActiveConnections()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+ConnectionPool::Cancellable* ConnectivityGrid::newStream(Http::ResponseDecoder& decoder,
+                                                         ConnectionPool::Callbacks& callbacks) {
+  if (pools_.empty()) {
+    createNextPool();
+  }
+
+  // TODO(#15649) track pools with successful connections: don't always start at
+  // the front of the list.
+  // The WrapperCallbacks is self owning, and will delete on pool success, failure, or cancellation.
```

Comment:
Self-owning objects scare me quite a bit, as they can easily lead to use-after-free. Is there no object nearby that could own this?